lib/util/source-code.js

```diff
@@ -301,32 +301,38 @@ class SourceCode extends TokenStore {
      */
     getJSDocComment(node) {
         let parent = node.parent;
-        const leadingComments = this.getComments(node).leading;
+        const leadingComments = this.getCommentsBefore(node);
 
         switch (node.type) {
             case "ClassDeclaration":
             case "FunctionDeclaration":
                 if (looksLikeExport(parent)) {
-                    return findJSDocComment(this.getComments(parent).leading, parent.loc.start.line);
+                    return findJSDocComment(this.getCommentsBefore(parent), parent.loc.start.line);
                 }
                 return findJSDocComment(leadingComments, node.loc.start.line);
 
             case "ClassExpression":
-                return findJSDocComment(this.getComments(parent.parent).leading, parent.parent.loc.start.line);
+                return findJSDocComment(this.getCommentsBefore(parent.parent), parent.parent.loc.start.line);
 
             case "ArrowFunctionExpression":
             case "FunctionExpression":
                 if (parent.type !== "CallExpression" && parent.type !== "NewExpression") {
-                    let parentLeadingComments = this.getComments(parent).leading;
-
-                    while (!parentLeadingComments.length && !/Function/.test(parent.type) && parent.type !== "MethodDefinition" && parent.type !== "Property") {
+                    let parentLeadingComments = this.getCommentsBefore(parent);
+
+                    while (
+                        !parentLeadingComments.length &&
+                        !/Function/.test(parent.type) &&
+                        parent.type !== "MethodDefinition" &&
+                        parent.type !== "Property" &&
+                        parent.type !== "Program"
```

Comment:
Calling `TokenStore.getCommentsBefore()` with a "Program" node will return comments that occur at the top of the file (this does not occur with `sourceCode.getComments()`, as it doesn't attach comments for Programs that have child nodes). This check is now required to avoid the the comment check bubbling up to the top block comment.